core/src/main/java/smile/manifold/UMAP.java

```diff
@@ -0,0 +1,846 @@
+     */
+    private static double[] findABParams(double spread, double minDist) {
+        int size = 300;
+        double[] x = new double[size];
+        double[] y = new double[size];
+        double xEnd = 3 * spread;
+        double xInterval = xEnd / (size - 1);
+        for (int i = 0; i < x.length; i++) {
+            x[i] = (i + 1) * xInterval;
+            y[i] = x[i] < minDist ? 1 : Math.exp(-(x[i] - minDist) / spread);
+        }
+        double[] p = {0.5, 0.0};
+        LevenbergMarquardt curveFit = LevenbergMarquardt.fit(func, x, y, p);
+        return curveFit.p;
+    }
+
+    /**
+     * Given a set of nearest neighbors and a neighborhood size, the fuzzy
+     * simplicial set (here represented as a fuzzy graph in the form of a sparse
+     * matrix) associated to the data. This is done by locally approximating
+     * geodesic distance at each point, creating a fuzzy simplicial set for each
+     * such point, and then combining all the local fuzzy simplicial sets into a
+     * global one via a fuzzy union.
+     * 
+     * @param knns
+     *            n nearest points for each data point
+     * @param n
+     *            The number of neighbors to consider when approximating the
+     *            neighbors.
+     * @param sigmas
+     *            Smooth approximator to knn-distance
+     * @param rhos
+     *            Distance to nearest neighbor
+     * @return A fuzzy simplicial set represented as a sparse matrix. The (i, j)
+     *         entry of the matrix represents the membership strength of the
+     *         1-simplex between the ith and jth sample points.
+     */
+    private static SparseMatrix fuzzySimplicialSet(Graph knns, int n, double[] sigmas, double[] rhos) {
+        smoothKnnDist(knns, n, sigmas, rhos, 64);
+        SparseMatrix ret = computeMembershipStrengths(knns, sigmas, rhos);
+        return ret;
+    }
+
+    /**
+     * Compute a continuous version of the distance to the kth nearest neighbor.
+     * That is, this is similar to knn-distance but allows continuous k values
+     * rather than requiring an integral k. In essence we are simply computing
+     * the distance such that the cardinality of fuzzy set we generate is k.
+     * 
+     * @param knns
+     *            n nearest points for each data point
+     * @param n
+     *            The number of neighbors to consider when approximating the
+     *            neighbors.
+     * @param sigmas
+     *            Smooth approximator to knn-distance
+     * @param rhos
+     *            Distance to nearest neighbor
+     * @param iterations
+     *            We need to binary search for the correct distance value. This
+     *            is the max number of iterations to use in such a search.
+     *            default 64
+     */
+    private static void smoothKnnDist(Graph knns, int n, double[] sigmas, double[] rhos, int iterations) {
+
+        double target = MathEx.log2(n);
+        Arrays.setAll(sigmas, sigmai -> 0.0);
```

Comment:
`Arrays.fill()` is better since you set a constant value here. Besides, you don't need this at all as sigma and rhos are initialized to 0.0 by default in Java.